File: Heap/2462. Total Cost to Hire K Workers/solution.py
# [17,12,10,2,7,2,11,20,8], k = 3, candidates = 4
class Solution:
    def totalCost(self, costs: List[int], k: int, candidates: int) -> int:
        minheap = []
        for i, cost in enumerate(costs):
            if i < candidates or i >= len(costs)-candidates:
                heapq.heappush(minheap, [cost, i])
        
        # keep tracking if any left worker still not push into the min heap
        l, r = candidates-1, len(costs)-candidates

        res = 0
        while k > 0:
            cost, idx = heapq.heappop(minheap)
            if l+1 < r and idx <= l:
                l+=1
                heapq.heappush(minheap, [costs[l], l])
            if l < r-1 and idx >= r:
                r-=1
                heapq.heappush(minheap, [costs[r], r])

            res += cost
            k -= 1
            
        return res

File: Heap/2462. Total Cost to Hire K Workers/solution.cpp
using pii = pair<int,int>;

class Solution {
public:
    long long totalCost(vector<int>& costs, int k, int candidates) {
        /*
        Core Logic: "Should a have LOWER priority than b?"
        The compare function returns true if the first element should have lower priority than the second element.
        Return Values:

        return true → a has lower priority than b
        return false → a has higher or equal priority than b
        */
        auto cmp = [](pii a, pii b) {
            if (a.first == b.first) {
                return a.second > b.second;
            }
            return a.first > b.first;
        };
        priority_queue<pii, vector<pii>, decltype(cmp)> min_heap; // or: priority_queue<pii, vector<pii>, greater<pii>> min_heap;
        int l=0, r=costs.size()-1;
        for (int i = 0; i < candidates && l <= r; i++) {
            min_heap.push({costs[l], l});
            l++;
        }
        for (int i = 0; i < candidates && l <= r; i++) {
            min_heap.push({costs[r], r});
            r--;
        }
        
        long long res = 0;
        for (int i=0; i<k; i++) {
            int cost = min_heap.top().first;
            int idx = min_heap.top().second;
            min_heap.pop();
            res += cost;
            
            if (l <= r) {
                if (idx < l) {
                    min_heap.push({costs[l], l});
                    l++;
                } else if (idx > r) {
                    min_heap.push({costs[r], r});
                    r--;
                }
            }
        }
        return res;
    }
};